register-machines/src/main.rs

```diff
@@ -11,55 +11,203 @@ enum Instruction {
+    fn encode(&self) -> Godel {
+        let (x, y) = self;
+        (1 << x.encode()) * (2 * y.encode() + 1) - 1
+    }
 }
-// a = (2^x)*(2y+1)-1
-fn decode_pair2(a: u128) -> (u128, u128) {
-    unimplemented!();
+
+impl<T: Encodable> Encodable for Option<T> {
+    fn encode(&self) -> Godel {
+        match self {
+            Some(x) => 1 + x.encode(),
+            None => 0
+        }
+    }
 }
-fn decode_godel_to_list(g: u128) -> Vec<u128> {
-    unimplemented!();
+
+impl Encodable for Instruction {
+    fn encode(&self) -> Godel {
+        match *self {
+            Add(i, j) => Some((2 * i, j)).encode(),
+            Sub(i, j, k) => Some((2 * i + 1, (j, k))).encode(),
+            Halt => (None as Option<u128>).encode()
+        }
+    }
+}
+
+impl<T: Encodable> Encodable for [T] {
+    fn encode(&self) -> Godel {
+        self.iter()
+            .rev()
+            .fold(
+                (None as Option<u128>).encode(),
+                |x, y| Some((y, x)).encode()
+            )
+    }
+}
+
+trait Decodable {
+    fn decode(godel: Godel) -> Self;
+}
+
+impl Decodable for u128 {
+    fn decode(godel: Godel) -> Self {
+        godel
+    }
+}
+
+impl Decodable for usize {
+    fn decode(godel: Godel) -> Self {
+        godel as usize
+    }
 }
+
+impl<T: Decodable, V: Decodable> Decodable for (T, V) {
+    fn decode(godel: Godel) -> Self {
+        let x_godel = godel.trailing_ones() as Godel;
+        let y_godel = godel >> (x_godel + 1);
+
+        (decode(x_godel), decode(y_godel))
+    }
+}
+
+impl<T: Decodable> Decodable for Option<T> {
+    fn decode(godel: Godel) -> Self {
+        if godel == 0 {
+            None
+        } else {
+            Some(decode(godel - 1))
+        }
+    }
+}
+
+impl Decodable for Instruction {
+    fn decode(godel: Godel) -> Self {
+        match decode::<Option<(u128, Godel)>>(godel) {
+            None => Halt,
+            Some((f, arg)) => {
+                if f % 2 == 0 {
+                    // EVEN: add
+                    Add(f / 2, decode(arg))
+                } else {
+                    // ODD: sub
+                    let (j, k) = decode(arg);
+                    Sub((f - 1) / 2, j, k)
+                }
+            }
+        }
+    }
+}
+
+impl<T: Decodable> Decodable for Vec<T> {
+    fn decode(mut godel: Godel) -> Self {
+        let mut list = vec!();
+
+        while let Some((head, tail)) = decode(godel) {
+            list.push(head);
+
+            godel = tail;
+        }
+    
+        list
+    }
+}
+
+fn decode<T: Decodable>(godel: Godel) -> T {
+    T::decode(godel)
+}
+
+use Instruction::*;
+
+fn eval_program(program: &[Instruction], init: &State) -> State {
+    let (mut label, mut registers) = init.clone();
+
+    while label < program.len() {
+        /* Reveal effects of instruction to state. */
+        match program[label] {
+            Add(reg, l) => {
+                /* Increment target register. */
+                *registers.get_mut(&reg).unwrap() += 1;
```

Comment:
This does work as long as you've properly defined the initial states of all variables, but will instantly panic if a user attempts to access a register not existing in the state. Therefore `*(registers.entry(&reg).or_insert(0)) += 1;` is seen as more idiomatic rust.